File: ws_ctrl.js
// ws_ctrl.js - Forward JS events and actions over WebSocket

const WsCtrl = (function() {

    // Flags to synchonize sending "init" event
    let isDomReady = false;
    let isWsOpen = false;

    let socket;
    let eventTrascoders = [];
    let debug = true;


    /**
     * Assigns a value to a (nested) property of a GLOBAL variable, or variable itself.
     *
     * @param {string} fullPath - The complete property path (e.g., 'DATA[0].array[INDEX_MAP.item]').
     * @param {*} value - The value to assign.
     */
    function dynaAssign(lhs, value) {
        let parentPath;
        let finalKey; // The actual key used for assignment

        // 1. Check the last character to determine the assignment type.
        const lastChar = lhs[lhs.length - 1];

        if (lastChar === ']') {
            // Array access case (requires eval on index expr)

            let bracketCount = 0;
            let splitIndex = -1;

            // Iterate backwards to find the matching opening bracket for the final expression.
            for (let i = lhs.length - 1; i >= 0; i--) {
                const char = lhs[i];

                if (char === ']') {
                    bracketCount++;
                } else if (char === '[') {
                    bracketCount--;
                    if (bracketCount === 0) {
                        // Found the start of the final array access expression.
                        splitIndex = i;
                        break;
                    }
                }
            }

            // Final key expression: Content inside the brackets (e.g., 'INDEX_MAP.item')
            const finalKeyExpression = lhs.substring(splitIndex + 1, lhs.length - 1);
            finalKey = globalThis.eval(finalKeyExpression);

            // Parent path: Everything before the final array access.
            parentPath = lhs.substring(0, splitIndex);

        } else {
            // Property access case (literal key, no eval needed for key)

            const lastDot = lhs.lastIndexOf('.');
            if (lastDot === -1) {
                // No path, direct assignment to global variable (e.g., 'DATA').
                globalThis[lhs] = value;
                return;
            }

            // Parent path: Everything up to the last dot.
            parentPath = lhs.substring(0, lastDot);

            // Final key: Literal key after the last dot. No eval needed.
            finalKey = lhs.substring(lastDot + 1);
        }

        // 3. EVAL #1 (Always Required): Get the reference to the immediate parent object.
        const parentObject = globalThis.eval(parentPath);

        // 4. Final assignment using the resolved parent reference and key.
        parentObject[finalKey] = value;
    }


    // Pass e.g. 'ws://localhost:8080'
    function init(ws_url) {
        socket = new WebSocket(ws_url);

        socket.onopen = (event) => {
            console.log("WsCtrl: WebSocket connected to " + ws_url);
            isWsOpen = true;
            sendInitEvent();
        };

        socket.onmessage = (event) => {
            if (debug)
                console.log('WsCtrl: Received: ', event.data);

            if (event.data.startsWith("{")) {
                const data = JSON.parse(event.data);
                switch (data.msg_type) {
                    case 'listen': {
                        eval(data.obj)[data.method](data.event_type, sendEvent);
                        break;
                    }
                    case 'call': {
                        const evaledObj = eval(data.obj);
                        const evaluatedArgs = data.args.map(arg => eval(arg));
                        const res = evaledObj[data.method](...evaluatedArgs);
                        if (debug)
                            console.log("WsCtrl: Call result:", res);
                        //const message = JSON.stringify({msg_type: "res", msg_id: data.msg_id, res: res});
                        //socket.send(message);
                        break;
                    }
                    case 'assign': {
                        dynaAssign(data.lhs, data.value)
                        break;
                    }
                    case 'exec': {
                        // eval expression and ignore result
                        eval(data.expr);
                        break;
                    }
                    case 'eval': {
                        // eval expression and send back result
                        const res = eval(data.expr);
                        sendMsg({msg_type: "res", msg_id: data.msg_id, res: res});
                        break;
                    }
                }
            }
        }

        socket.onerror = (error) => {
            console.error("WsCtrl: WebSocket Error:", error);
        }
    }


    function addEventTranscoder(f) {
        eventTrascoders.push(f);
    }


    function sendMsg(msg) {
        if (debug)
            console.log("WsCtrl: Sending:", msg)
        socket.send(JSON.stringify(msg));
    }


    function sendEvent(event) {
        if (debug)
            console.log("WsCtrl: sendEvent:", event)

        const edata = {
            msg_type: 'event',
            _class_: event.constructor.name,
            type: event.type,
            clientX: event.clientX,
            clientY: event.clientY,
        };

        if (typeof event.target === 'object') {
            edata.targetId = event.target.id;
            edata.targetTag = event.target.tagName;
            edata.value = event.target.value;
        }

        for (const f of eventTrascoders) {
            f.call(null, event, edata);
        }

        if (socket.readyState === WebSocket.OPEN) {
            sendMsg(edata);
        } else {
            console.warn("WsCtrl: WebSocket not open, cannot send event:", event.type);
        }
    }


    function sendInitEvent() {
        // Send event only when both DOM and websockets are ready
        if (isDomReady && isWsOpen) {
            sendMsg({msg_type: "init"});
        }
    }


    function setDebug(flag) {
        debug = flag;
    }


    document.addEventListener('DOMContentLoaded', (event) => {
        isDomReady = true;
        sendInitEvent();
    });


    return {
        init: init,
        addEventTranscoder: addEventTranscoder,
        setDebug: setDebug,
    }

})();

File: ws_ctrl.py
import asyncio
import websockets
import json
import logging


class JWS:

    def __init__(self, ws):
        self.ws = ws
        self.msg_id = 1
        self.futures = {}

    @staticmethod
    def map_arg(arg):
        if isinstance(arg, JSExpr):
            arg = str(arg)
        elif isinstance(arg, str):
            arg = "'%s'" % arg.replace("'", "\\'")
        return arg

    async def msg(self, msg):
        msg["msg_id"] = self.msg_id
        self.msg_id += 1
        logging.debug("Sending: %s", msg)
        await self.ws.send(json.dumps(msg))
        return msg["msg_id"]

    async def listen(self, obj, event_type, method="addEventListener"):
        await self.msg({"msg_type": "listen", "method": method, "obj": self.map_arg(obj), "event_type": event_type})

    async def call(self, obj, method, args):
        outargs = [self.map_arg(arg) for arg in args]
        msg = {"msg_type": "call", "obj": self.map_arg(obj), "method": method, "args": outargs}
        return await self.msg(msg)

    async def exec(self, expr):
        msg = {"msg_type": "exec", "expr": expr}
        return await self.msg(msg)

    async def eval(self, expr):
        msg = {"msg_type": "eval", "expr": expr}
        msg_id = await self.msg(msg)
        return await self.future_result(msg_id)

    async def assign(self, lhs, value):
        msg = {"msg_type": "assign", "lhs": self.map_arg(lhs), "value": value}
        return await self.msg(msg)

    async def future_result(self, msg_id):
        fut = asyncio.Future()
        self.futures[msg_id] = fut
        return await fut


class JSExpr:

    def __init__(self, e=None, jws=None):
        self.__e = e
        self.__jws = jws

    def __getattr__(self, attr):
        if self.__e is None:
            e = JSExpr(attr, self.__jws)
            return e
        else:
            self.__e += "." + attr
            return self

    async def __call__(self, *args):
        arr = str(self).rsplit(".", 1)
        if len(arr) == 1:
            obj = None
            meth = arr[0]
        else:
            obj, meth = arr
            obj = JSExpr(obj)
        return await self.__jws.call(obj, meth, args)

    def __str__(self):
        return self.__e

    def __repr__(self):
        return "JSExpr(%s)" % self.__e



async def _websocket_handler(websocket, handle_event):
    """
    Handles incoming WebSocket connections and messages.
    """
    logging.info(f"Client connected from {websocket.remote_address}")

    jws = JWS(websocket)
    js = JSExpr(None, jws)

    try:
        async for message in websocket:
            logging.debug(f"Received: {message}")
            try:
                event = json.loads(message)
            except json.JSONDecodeError:
                logging.error(f"Received non-JSON message: {message}, ignoring")
                continue

            if event["msg_type"] == "res":
                fut = jws.futures.pop(event["msg_id"], None)
                if fut is None:
                    logging.warning("Received result for msg_id=%d, but no future registered", event["msg_id"])
                else:
                    fut.set_result(event["res"])
            else:
                asyncio.create_task(handle_event(jws, js, event))

    except websockets.exceptions.ConnectionClosed as e:
        logging.info(f"Client disconnected: {e.code} - {e.reason}")
    except Exception as e:
        logging.exception(f"An unexpected error occurred: {e}")
    finally:
        logging.info(f"Connection closed with {websocket.remote_address}")


async def ws_server(user_handler, host="localhost", port=8080):
    async with websockets.serve(lambda a: _websocket_handler(a, user_handler), host, port):
        logging.info("WebSocket server started on ws://%s:%d" % (host, port))
        await asyncio.Future()  # Run forever


if __name__ == "__main__":

    async def my_handler(jws, js, event):
        print(event)
        if event["msg_type"] == "init":
            # Handle "init" event here and install listeners for needed JS events
            await jws.listen(js.document.body, "click")

    asyncio.run(ws_server(my_handler))
